Adds auto dependencies for make

ksschwabe opened the pull request:
This resolves the issue (#77) that if any of the header header files were
modified, make would not recompile to include the changes in the header
files the unless clean was run before recompiling.

.o object files and .d dependency files are created for each of the c files. The only proviso is that if a header file is renamed, the dependency will be "lost" so you will have to `make clean` before recompiling.

ksschwabe commented:
@davids5 : FYI
@dagar: I know you assigned this to yourself, but perhaps this is what you wanted anyway.

I am not very good with the complexities of Makefiles, so please check  that I haven't done anything obviously silly with the makefile.

dagar commented:
Thanks! I had forgotten about this.

ksschwabe commented:
I see the Travis-CI build has failed because of ROM overflow issues. I am not sure why, though....

Master seems to build, but my PR doesn't; I haven't changed any optimisations. It might have to do with how the elf is created if it is created directly from SRCS vs. linking object files.....

ksschwabe commented on Makefile.f4:
```diff
@@ -25,8 +28,16 @@ BINARY		 = $(TARGET_FILE_NAME).bin
+# Compile and generate dependency files
+$(OBJS):	$(SRCS)
+	$(CC) -c $(FLAGS) $*.c -o $*.o
+	$(CC) -MM $(FLAGS) $*.c > $*.d
```
These two lines should be checked to see if they can be improved to to be more generic. I don't like that it is hard coded that the c-files become o-files.
```
        $(CC) -c $(FLAGS) $*.c -o $*.o
        $(CC) -MM $(FLAGS) $*.c > $*.d
```

ksschwabe commented:
I have changed this to WIP.

If I make the pxio_bl independently using `make pxio_bl`, it compiles correctly and doesn't have any ROM overflow issues. But if I compile it after other files, like in `make all`, then the dependencies and objects are "already built" and are not updated, even though they should be updated for the different compilation defines, so it then overflows in ROM. ... <Sigh> 

Any suggestions? Does this mean I have to create a "Build directory" for each board to avoid this problem?

davids5 commented:
Yes you will need a separate build dir per board. Because a files xxx.o meets the dependency so it is not rebuild, but the the build flags have changed, in this case F4 related to the floating point settings that are different than F1 ones.

@dagar is it time for cmake? 

If not, since the changes are all the same, how about creating a rules.mk and including it in the Makefile.xx files.
Also ```-include $(DEPS)``` should be at the bottom of the file.

ksschwabe commented on rules.mk:
```diff
@@ -0,0 +1,32 @@
+all:		$(BUILD_DIR) $(ELF) $(BINARY)
+
+# Compile and generate dependency files
+$(BUILD_DIR)/%.o:	%.c
```
@davids5 : Is there a more elegant way to handle the dependency on the particular source file so that you don't have to specifically mention its file extension?

davids5 replied:
I am not sure what you are asking. You are already using the pattern rule and define OBJS a list conversion .c -> .o . What more did you want to do?

ksschwabe replied:
It was more a question of style, and whether this is how it should be done.

davids5 replied:
I think it is fine. Since the rule is %.o: %.c the $*.c is appropriate

ksschwabe commented:
@davids5 
After quite a struggle and getting to grips with how Make works, I seem to have something that now works for the purposes of compiling and using dependencies, so if a header file is touched, it will rebuild the respective object files.

I know there is a Travis-CI error relating to it not being able to run the `deploy`  target, because the .bin files are now located in the individual build directories, but I wanted your feedback before I fix that.

1. Should I leave the binaries and .elf files in the build directories, or should I generate them in the root bootloader directory instead and only put the object and dependency files in the build directories?
2. Should I generate the dependency files with the -MMD or the -MD option (-MD includes system headers; -MMD doesn't)? Does it make a difference to us here?

dagar commented:
I'm willing to add a cmake build for this if you guys think it's worth it and want to validate.

davids5 commented:
@dagar The makes sense me.

ksschwabe commented:
Fixed `make deploy` CI issue. `make deploy` collects all of the .bin files in the build_* directories and puts them in to a single Bootloader.zip.

ksschwabe commented:
@davids5 : Is there anything else that you would like me to do with the makefile? Or are there any further recommendations that you have for making the makefile more readable or more easily extendible in the future?

davids5 commented:
@ksschwabe 

Thank you!

You could add `sizes` target.  Then invoke it as the last step of all. See Firmware master. Once you are done let me know and I will pull and test this.

ksschwabe commented:
@davids5: I have added the `sizes` target, which is called as the last step of 'all'. When you have a chance, you can pull and test it.

davids5 commented on Makefile:
```diff
@@ -58,7 +58,7 @@ all:	$(TARGETS)
 clean:
 	cd libopencm3 && make --no-print-directory clean && cd ..
-	rm -f *.elf *.bin
+	rm -r build_*
```
type `make clean` 2 times. 
On the second invocation you will get (or on the first on fresh pull)
```
rm: cannot remove ‘build_*’: No such file or directory
make: *** [clean] Error 1
```
To have `make clean` not issue the error, change `-r` -fr`

davids5 replied:
I was thinking that we can cut down on some of the build noise

You could do something like:
[0007-Cut-down-on-noise.patch.txt](https://github.com/PX4/Bootloader/files/1122616/0007-Cut-down-on-noise.patch.txt)

ksschwabe commented:
@davids5 : I made the two changes, and applied your patch to reduce the build noise. I also changed the last remaining `make` to a `${MAKE}` in the Makefile, to be consistent.

One question: where is the variable `${MAKE}` defined? I see that it is evaluated to `make`, but I can't find where the original definition is coming from. Is the idea that this is a system variable that you can override with a different make version like `colormake` if you wanted to?

davids5 commented:
>One question: where is the variable ${MAKE} defined? I see that it is evaluated to make, but I can't find where the original definition is coming from. Is the idea that this is a system variable that you can override with a different make version like colormake if you wanted to?

https://www.gnu.org/software/make/manual/html_node/MAKE-Variable.html#MAKE-Variable

davids5 commented:
@ksschwabe - 
There is one more thing. If you check out master then build it. Then check out this branch and build it 
you will get 
```
rm-none-eabi-objcopy -O binary build_tapv1_bl/tapv1_bl.elf build_tapv1_bl/tapv1_bl.bin
find: paths must precede expression: auavx2v1_bl.elf
Usage: find [-H] [-L] [-P] [-Olevel] [-D help|tree|search|stat|rates|opt|exec] [path...] [expression]
```
but `make clean` will not fix it.  

 `rm -f *.bin *.elf` will fix it.

I think we need to fix the `sizes` target, but I think folks may get the "Hey my fix is not working syndrome" If we do not purge the old build files.

We could add *.bin *.elf back in `clean` OR we can make a `distclean` target that calls clean and have it and add `rm -f *.bin *.elf` 

I can lean both ways on this:

Thinking: What if someone wnats to keep a bin in the root for A/B testing. `distclean` makes the most sense.  BUT We will have to document the `make distclean` in the ReadME or the average builder could  suffer. 

....Ah put it in the clean. The advanced uses can add a folder for the old bins.

davids5 commented:
@ksschwabe the fix to find is:

```
-       @-find build_* -name *.elf -type f | xargs size 2> /dev/null || :
+       @-find build_* -name '*.elf' -type f | xargs size 2> /dev/null || :
```

ksschwabe commented:
@davids5 : Sorry for the delay in fixing this; last week became a bit chaotic.

I have added your fix for the `find: paths must precede expression` error when building in a directory that already contains elf files, namely:
```
    -       @-find build_* -name *.elf -type f | xargs size 2> /dev/null || :
    +       @-find build_* -name '*.elf' -type f | xargs size 2> /dev/null || :
```

I have also added the `rm -f *.bin *.elf` line back to the `clean` target.

davids5 commented:
@ksschwabe Thank you!
